Migration update

Filienko opened the pull request:
Migration to generate migration files and then iterate over them to make sure that the state of the system is at the latest possible resource configuration

Filienko committed:
WIP: starting migration work

Filienko committed:
WIP: run_migration CLI

Filienko committed:
Moving the migration into separate module

pbugni commented on isacc_messaging/api/views.py:
```diff
@@ -2,6 +2,7 @@
 import click
 import logging
 from datetime import datetime
+import os
```
i think this isn't needed, now that you moved the migration code to a separate module

pbugni commented on migration/migrations/001_initial.py:
```diff
@@ -0,0 +1 @@
+print("Initial Migration")
```
great start!
you'll certainly want to capture the `revision` string and `down_revision` (none in the initial case, only), for the migration framework to determine the order, etc.

pbugni replied:
please nest all the migration scripts in another directory, i.e. `migration/versions/*.py`.  this will make it easier on the Migration class to find and sort them.

pbugni commented on migration/migration.py:
```diff
@@ -0,0 +1,102 @@
+def downgrade():
+    # Run migrations to downgrade the schema
+    run_migrations("downgrade")
+
```
I'd further break this up.  use `migration/views.py` for all the flask view functions, and all the migration logic can remain in `migration/migration.py`

pbugni reviewed (commented):
great progress

Filienko committed:
WIP: migration class

Filienko committed:
WIP: separating migratory functionality

Filienko committed:
WIP: fixing naming

pbugni commented on isacc_messaging/api/views.py:
```diff
@@ -302,3 +304,22 @@ def deactivate_patient(patient_id):
         f"Patient {patient_id} active set to false",
         level='info'
     )
+
```
everything below here should move to /migration/views.py - get its own blueprint, and be loaded in with the base_blueprint in `app.py`

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,139 @@
+
+    def get_migration_files(self) -> List[str]:
+        """Retrieve the list of migration files."""
+        migration_files = sorted(os.listdir(self.migrations_dir))
```
great progress, i like where this is going.

i believe you'll need to look for the `revision` and `down_revision` numbers in the files to get the order, confirm no cycles, etc.

pbugni replied:
probably worth adding a few mock test files, to confirm your `Migration.migration_files` end up ordered respecting the `down_revision` reverse order.  for bonus points, confirm it errors if there's a gap or cycle.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,139 @@
+        if basic is None:
+            self.create_new_migration_manager()
+
+            return 1
```
i'd return `None` here.  we haven't yet successfully upgraded to this version number you are returning, right?

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,139 @@
+
+        return latest_applied_migration
+
+    def update_applied_migration_in_fhir(self, latest_applied_migration: int):
```
i probably confused the issue, using simple integers when discussing the examples for how to order.  the problem with using integers is there's no flexibility in reordering - say two dev branches grow apart and then come back together with batches of upgrade steps in each.  you wouldn't want to hand adjust a whole batch of revision and back_revision values.

a random hash string as the revision number solves this.  the order is just defined by the respective `down_revision` values in the scripts.  to reorder a fork, only a single down_revision edit is needed.

Filienko replied:
Thank you! I was slightly confused, sorry.

pbugni commented on migrations/migration_resource.py:
```diff
@@ -0,0 +1,37 @@
+        return f"{self.resource_type}/{self.id}"
+
+    def get_latest_migration():
+        return 0
```
I imagine this is just a place holder, but again, I'd return `None` if there isn't one already persisted.

Filienko committed:
Refactoring and adding mock tests

Filienko committed:
Modifying tests

Filienko committed:
WIP: fixing tests

Filienko committed:
WIP: fixing tests

Filienko committed:
WIP: tests

Filienko committed:
WIP: fixing tests

Filienko committed:
WIP: fixing test

Filienko committed:
Fixing tests

Filienko committed:
WIP: fixing tests

Filienko committed:
WIP: fixing errors

Filienko committed:
Fixing formatting

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,178 @@
+
+        def dfs(node):
+            if node in traversing:
+                error_message = f"Cycle detected in migration sequence for {node}"
```
very nice - thanks for testing for cycles.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,178 @@
+        created_time = datetime.now().astimezone().isoformat()
+        m = {
+            'resourceType': 'Basic',
+            'identifier': [{"system": "http://our.migration.system", "value": system_wide_id_or_latest_applied_migration}],
```
i think `identifier` is overkill here - duplicate data gets out of sync if we don't need it, lets not.

pbugni commented on migrations/migration_resource.py:
```diff
@@ -0,0 +1,45 @@
+"""Migration Module
+
+Captures common methods needed by the system for migrations, by specializing
+the `fhirclient.Basic` class.
```
Add something to explain how we use, i.e.:

A single `Basic` resource is maintained with the most recently (successfully) run migration revision held in the single `Basic.code` value.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,178 @@
+        with open(migration_path, "r") as migration_file:
+            migration_code = migration_file.read()
+            try:
+                exec(migration_code)
```
`exec` is notoriously dangerous.  we should just use entry points within the file.  please try `__import__(migration_file)` or see the `imp` module which you can use to import the migration file by path.  Then call `upgrade` or `downgrade` directly.

pbugni commented on isacc_messaging/api/views.py:
```diff
@@ -302,3 +301,4 @@ def deactivate_patient(patient_id):
         f"Patient {patient_id} active set to false",
         level='info'
     )
+
```
whitespace only changes - may as well revert these for a clean PR.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,178 @@
+            migration_file.write(f"# Current version: {new_id}\n")
+            migration_file.write(f"# Previous version: {current_id}\n")
+            migration_file.write("\n")
+            migration_file.write("# Add your migration code here\n")
```
please add boilerplate for `upgrade()` and `downgrade()` functions, to be called respective of which direction the user is moving.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,178 @@
+        basic = first_in_bundle(basic)
+
+        if basic is None:
+            self.create_applied_migration_manager()
```
i wouldn't create it, until we've run a successful migration.  just return None as you are doing here.

pbugni commented on migrations/migration_resource.py:
```diff
@@ -0,0 +1,45 @@
+    def update_migration(self, migration_id):
+        """Persist self state to FHIR store"""
+        self.code = {"coding": [{ "system": "http://our.migration.system", "code": migration_id}]}
+        response = self.persist()
```
if your intent with `created` was to track when the migration was run, you'll need to update here as well.

pbugni commented on migrations/migration_resource.py:
```diff
@@ -0,0 +1,45 @@
+        response = self.persist()
+        return response
+    
+    def get_migration(self):
```
name this function to explain which migration we're getting.  is it returning the last one run?  where did `self.code` get initialized?

pbugni commented on migrations/versions/test_migration_0.py:
```diff
@@ -0,0 +1,4 @@
+# Migration script generated for test_migration_0
+# Current version: 0
```
best to test what we're using.  generate a random `uuid.uuid4()` and save in your test_migration file for testing, rather than simple ints, to avoid surprises.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,178 @@
+        migration_path = os.path.join(self.migrations_dir, migration_filename)
+        with open(migration_path, "w") as migration_file:
+            migration_file.write(f"# Migration script generated for {migration_name}\n")
+            migration_file.write(f"# Current version: {new_id}\n")
```
it would probably be easier to us variables rather than comments for `revision` and `down_revision`, then you can let python do the parsing via `import`.

Filienko replied:
Should I use imp or is there a better practice?

Filienko committed:
Addressing comments

Filienko committed:
Tests

Filienko committed:
WIP: fixing tests

Filienko committed:
WIP: fixing test

Filienko committed:
Fixing tests

Filienko committed:
WIP: working single tree

Filienko committed:
Changed the migration from referencing the latest applied to the latest created file
(Reversable)

Filienko committed:
Only allow migreation to be generated, if the system is caught up

Filienko committed:
WIP: adding testing

Filienko committed:
Removing tests

Filienko committed:
Adding back the test case

Filienko committed:
Fixing upgrade

Filienko committed:
Fixing a test

Filienko committed:
Changed to linked list

Filienko committed:
WIP: adding test

Filienko committed:
WIP: testing

Filienko committed:
WIP: testing

Filienko committed:
WIP: testing

Filienko committed:
Fixing cycles

Filienko committed:
WIP: adding eq

Filienko committed:
Fixing cycles

Filienko committed:
WIP: fixing cycles

Filienko committed:
WIP: fixing migration

Filienko committed:
WIP: fixing

Filienko committed:
WIP: fixing migration

Filienko committed:
WIP: fixing migration

Filienko committed:
Testing cycle

Filienko committed:
errors

Filienko committed:
Fixing errors

Filienko committed:
errors

Filienko committed:
errors

Filienko committed:
Refactoring Linked List utilities

Filienko committed:
WIP: fixing cycles

Filienko committed:
WIP: fixing cycles

Filienko committed:
Fixing cycles

Filienko committed:
Fixing data

Filienko committed:
WIP:fix

Filienko committed:
Testing

Filienko committed:
WIP: fixing errors

Filienko committed:
WIP: new

Filienko committed:
WIP: fixing errors

Filienko committed:
WIP: fixing cycle

Filienko committed:
WIP: fixing errors

Filienko committed:
Fixing errors

Filienko committed:
WIP: fixing erros

Filienko committed:
WIp: fixing

Filienko committed:
fixing

Filienko committed:
fixing circular

Filienko committed:
Adding tests

Filienko committed:
Changing logic

Filienko committed:
utils

Filienko committed:
fix list fixture

Filienko committed:
fix

Filienko committed:
Fixing test

pbugni commented on migrations/views.py:
```diff
@@ -0,0 +1,39 @@
+    migration_manager.generate_migration_script(migration_name)
+
+
+@migration_blueprint.cli.command("insert")
```
this looks to be a duplicate of `migrate`

pbugni commented on migrations/views.py:
```diff
@@ -0,0 +1,39 @@
+    migration_manager.generate_migration_script(migration_name)
+
+
+@migration_blueprint.cli.command("upgrade")
```
please add help strings to all of these cli.commands (and arguments), so users can review that these functions do from the command line (via `<command_name> --help`)

pbugni commented on migrations/views.py:
```diff
@@ -0,0 +1,39 @@
+
+
+@migration_blueprint.cli.command("restart")
+def upgrade():
```
duplicate function name `upgrade` will fail.  i think `reset` may be a better name for the cli.command and function.

pbugni commented on tests/migration_list_test.py:
```diff
@@ -0,0 +1,80 @@
+    assert hash(node1) == hash("data1")
+    assert hash(node1) != hash(node2)
+    assert hash(node1) == hash(node3)
+
```
wonderful job with unit tests here.  💯

pbugni commented on tests/migration_test.py:
```diff
@@ -0,0 +1,132 @@
+    def raise_for_status(self):
+        if self.status_code == 200:
+            return
+        raise Exception("status code ain't 200")
```
would be better to mock the exception type and details raised by `requests.raise_for_status`

pbugni commented on tests/migration_test.py:
```diff
@@ -0,0 +1,132 @@
+    # Mock the output of get_migration_files
+    with patch.object(Migration, 'get_migration_files', return_value=[]):
+        # Instantiate Migration class
+        migration_instance = Migration()
```
puzzled to see `migration_instance` fixture, and calls to do the same?

Filienko replied:
Some of the methods are easier to test from clean slate, instead of importing existing Migration. Can alter.

pbugni commented on tests/migration_test.py:
```diff
@@ -0,0 +1,132 @@
+    assert down_revision == "migration122"
+
+    # Delete the migration file after assertion
+    os.remove(migration_path)
```
another great set of test! 💯

pbugni commented on tests/migration_test.py:
```diff
@@ -0,0 +1,132 @@
+def test_get_previous_migration_id_exists(migration_instance):
+    migration = "migration123"
+    migration_content = "down_revision = 'migration122'\n"
+    migration_path = os.path.join(migration_instance.migrations_dir, migration + ".py")
```
you probably don't want to be writing (or reading from) test files to the configured `migrations_dir`.  your tests should use a fake / dummy location.

pbugni commented on migrations/versions/test_7c929f8e-bd11-4283-9603-40613839d23a.py:
```diff
@@ -0,0 +1,5 @@
+# Migration script generated for test_migration_2
```
this should live somewhere in the tests directory

Filienko replied:
Will delete it and use mock instead

pbugni commented on migrations/utils.py:
```diff
@@ -0,0 +1,124 @@
+                return current_node
+            current_node = current_node.next_node
+
+        # If not found, start from the tail node
```
how can `self.head` point at either end, as implied by these two blocks?

pbugni commented on migrations/utils.py:
```diff
@@ -0,0 +1,124 @@
+    def append(self, data):
+        """Append node after the head."""
+        new_node = self.find(data)
+        # If do not exist, create the nodes
```
wouldn't it be an error, if the node already exists?   i believe the implementation here doesn't handle duplicates.  this should be documented in the class docstring.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,231 @@
+                error_message = "Cycle detected in migration sequence"
+                audit_entry(error_message, level='error')
+                raise ValueError(error_message)
+        else:
```
I don't believe you want to generate an error here.  We should be able to add the migration library to a system, without requiring any migration scripts, right?

pbugni replied:
never mind - just the variable name, not an error being raised.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,231 @@
+            try:
+                migration_module = imp.load_source("migration_module", migration_path)
+                down_revision = getattr(migration_module, "down_revision", None)
+            except Exception as e:
```
catching base `Exception` is dangerous.  it will often lead to other unexpected situations being caught and treated like the expected case.

please except the type of exception you expect to catch only.

also, in this case, I would think you'd want to re `raise e` as it's rather exceptional to find an ill defined migration file - and without re-rasing, the situation could easily be ignored.

Filienko committed:
Responding to comments

Filienko committed:
Fixing add

Filienko committed:
Adding comments

Filienko committed:
Fixing tests

Filienko committed:
Fixing tests

Filienko committed:
Tests

mcjustin commented:
@pbugni @ivan-c  just an FYI - I didn't get any responses to my query about a FHIR migration tool a couple weeks ago...
https://chat.fhir.org/#narrow/stream/179239-tooling/topic/Migration.20of.20resources.20-.20tooling.3F/near/435486203

Filienko committed:
Tests

Filienko committed:
Fixed Test

Filienko committed:
migration

Filienko committed:
Fixing tests

Filienko committed:
WIP: test

Filienko committed:
WIP: file

Filienko committed:
Fixing test

Filienko committed:
WIP: fixing

Filienko committed:
WIP: imp

Filienko committed:
WIP: imp

Filienko committed:
Removing .imp test

Filienko committed:
WIP: removing test

Filienko committed:
Adding property

Filienko committed:
Decoupling name from version

Filienko committed:
Enforce unique  migration names

Filienko committed:
Fixing test

Filienko committed:
Adding errors

Filienko committed:
Testing added

Filienko committed:
Fixing tests

Filienko committed:
Errors

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,264 @@
+                if revision:
+                    revisions.append(revision)
+                    self.migrations_locations[revision] = module_name
+            except Exception as e:
```
please catch just the expected Exception type.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,264 @@
+            unapplied_migrations = current_migration
+
+        if not unapplied_migrations or unapplied_migrations == 'None':
+            raise ValueError("No valid migration files to run.")
```
why is this an exceptional case?  (to not find any unapplied migrations)  i'd expect a silent run if attempting `upgrade` and finding the system already up to date.

Filienko replied:
Will remove.

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,264 @@
+    def get_latest_applied_migration_from_fhir(self) -> str:
+        """Retrieve the latest applied migration migration id from FHIR."""
+        # Logic to retrieve the latest applied migration number from FHIR
+        basic = MigrationManager.get_resource()
```
i see opportunity to DRY this out a bit.  both this method and the next have a half dozen lines of code that could be done once in `MigrationManager.get_resource()`  just needs a flag to know if it should `create_if_not_found`

pbugni commented on migrations/migration.py:
```diff
@@ -0,0 +1,264 @@
+
+
+    def create_applied_migration_manager(self, initial_applied_migration: str = None):
+        """Create new FHIR resource to keep track of Migration History."""
```
this feels redundant - can't it be done within the `MigrationManager()` class?

pbugni commented on migrations/utils.py:
```diff
@@ -0,0 +1,167 @@
+        return hash(self._data)
+
+
+class LinkedList:
```
please add [docstrings](https://peps.python.org/pep-0257/) to all classes.  this one should include something like:

Specialized doubly linked list implementation restricted to unique data values and no cycles.

pbugni commented on migrations/utils.py:
```diff
@@ -0,0 +1,167 @@
+                self._head = new_node
+            else:
+                current_node = self._head
+                while current_node.next_node:
```
aren't we adding a single node after `self._head` on every call here?  the `while` seems unnecessary, wouldn't it be exceptional to find `self._head.next_node != None` ?

pbugni replied:
i don't see how this `while` loop is necessary.  doesn't `self._head` point to the end of the list? wouldn't it be an exception `if self._head.next_node is not None` ?

pbugni commented on migrations/utils.py:
```diff
@@ -0,0 +1,167 @@
+        new_node = self.find(data)
+        if new_node:
+            raise ValueError("adding a duplicate item")
+        else:
```
the else clause (and indentation) is unnecessary.  the "guard clause" above will exit prematurely if needed.

pbugni commented on migrations/utils.py:
```diff
@@ -0,0 +1,167 @@
+            error_message = "Cycle detected in the list"
+            raise ValueError(error_message)
+
+    @property
```
@properties are great, but i don't see the value of this one.  aren't you just wrapping what could be done directly?  unless there are additional details handled in the @property or @setter, it's un-pythonic to wrap just for the sake of wrapping.

Filienko committed:
New version

Filienko committed:
Adding tests

Filienko committed:
Fixing

Filienko committed:
Adding tests

Filienko committed:
Tests

Filienko committed:
Modifying tests

Filienko committed:
test

Filienko committed:
Fixing tests

pbugni reviewed (approved):
Thank you!  Looks great.

Filienko committed:
Added example

Filienko committed:
Small adjusments to the text

ivan-c commented on migrations/views.py:
```diff
@@ -0,0 +1,44 @@
+"""Migration flask commands
+
+Defines number of commands relevant for creating and managing migrations.
```
minor nitpick: I don't think anything in this file responds to HTTP requests, so `views.py` may be a little misleading. Unless you plan on adding routes later (so these can be invoked by HTTP), I might suggest renaming so it's more obvious (eg `commands.py`).

ivan-c commented on migrations/migration.py:
```diff
@@ -0,0 +1,254 @@
+In order to create a new migration, all existing migrations must
+first be applied.
+
+You cannot create mpre than one new migration file at a time.
```
```suggestion
You cannot create more than one new migration file at a time.
```

ivan-c commented on migrations/views.py:
```diff
@@ -0,0 +1,44 @@
+
+from migrations.migration import Migration
+
+migration_blueprint = Blueprint('migration', __name__, cli_group=None)
```
neat! wasn't aware flask integrated click commands directly into blueprints

https://flask.palletsprojects.com/en/2.3.x/cli/#registering-commands-with-blueprints

ivan-c commented on migrations/migration.py:
```diff
@@ -0,0 +1,254 @@
+                    revisions.append(revision)
+                    self.migrations_locations[revision] = module_name
+            except KeyError as e:
+                print(f"Error loading migration script {module_name}: {e}")
```
I haven't thought through all the ways this exception could be triggered, but it might be safer to exit on error (or just leave this exception unhandled until you see it first-hand)

ivan-c commented on migrations/migration.py:
```diff
@@ -0,0 +1,254 @@
+                migration_module = imp.load_source("migration_module", migration_path)
+                down_revision = getattr(migration_module, "down_revision", None)
+            except Exception as e:
+                message = f"Error loading migration script {filename}: {e}"
+                audit_entry(
+                    message,
+                    level='debug'
+                )
+                raise e
+        else:
+            message = f"Migration script {filename} does not exist."
+            audit_entry(
+                message,
+                level='debug'
+            )
```
If I understand correctly, loading the migration files happens well before they're applied. 

It's a minor nitpick, but since the data should be unaltered at this point, it may make more sense to use `logger.error` instead of adding an audit log record (generally added when data is modified)

Filienko replied:
What is the best way to configure it?

ivan-c commented on migrations/migration_resource.py:
```diff
@@ -0,0 +1,78 @@
+"""
+from datetime import datetime
+
+from fhirclient.models.basic import Basic
```
minor nitpick: `fhirclient` is pretty awful and would be an ongoing burden to maintain. there may be better alternatives like [fhir.resources](https://pypi.org/project/fhir.resources/)

the biggest issue is that they tightly couple the version of PyPI to the version of FHIR they support. ie `fhirclient==4.0.0` can **only** support FHIR R4

ivan-c commented on migrations/migration_resource.py:
```diff
@@ -0,0 +1,78 @@
+                'created': created_time,
+            }
+
+        response = HAPI_request('POST', 'Basic', resource=resource)
```
Although we should be able to easily search for our Basic resource given the system/code, I think it would be more consistent (and one fewer HTTP request) to PUT (upsert) a Basic resource to a consistent FHIR resource ID (eg `cirg-migration-version`)

ivan-c commented on migrations/utils.py:
```diff
@@ -0,0 +1,208 @@
+            current_node = current_node.prev_node
+        return None
+
+    def present(self, data) -> bool:
```
I think you could override `__contains__()` so you could use the `in` keyword to make it slightly more idiomatic.
I don't see this invoked anywhere, so it's only a nitpick

ivan-c commented on migrations/utils.py:
```diff
@@ -0,0 +1,208 @@
+        '''Iterates from the head to check whether all of the nodes were rightly assigned 
+        prev and next node references'''
+        
+        # If no tail node exists and length is not zero, means there is a circual dependency, no outgoing edges
```
```suggestion
        # If no tail node exists and length is not zero, means there is a circular dependency, no outgoing edges
```

ivan-c commented on migrations/example/versions/add_identifier.py:
```diff
@@ -0,0 +1,56 @@
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+
+# Define the FHIR server base URL
+FHIR_SERVER_URL = 'http://localhost:8080/fhir'
```
minor nitpick: this value will change depending on where the migration is run. it'd be nice to have an example where this is read from a config value (ie `current_app.config['FHIR_URL']`)

Filienko replied:
I thought we would want to limit involvement of outside libraries, i.e. flask, since we hope to put it into a library?

ivan-c reviewed (approved):
Looks good, thanks @Filienko!! Had a few nitpicks, but ready to merge

ivan-c commented:
@Filienko 

One minor request- since this PR has so many commits, when you merge, please select **Squash and merge** from the merge button :bow:

Filienko committed:
Addressing comments
WIP: put

Filienko committed:
Constant system comparison

pbugni committed:
Adding temporary function to test direct PUT that is failing from within
the code.

pbugni committed:
Left an exception where the implementation doesn't match the spec

pbugni committed:
Ported away from HAPI_request, preferring requests on this module.
Eliminated a number of repeated strings using common variables
Use logging rather than print
When including query string parameters, use built in function
parameters, as special characters need to be URL encoded.

Filienko committed:
Fixing

Filienko committed:
Removing test

Filienko committed:
Made migrations more robust

Filienko committed:
Removing isacc

Filienko committed:
Changing default number to uuid
